### PR TITLE
Adding Mock with CSI 0.2.0

### PR DESCRIPTION
**Pull request by sbezverk**
Until official GoCSI compatible with 0.2.0 spec changes is produced, I propose to add this modified `mock` driver which was refactored to support 0.2.0 changes.

### CODE REVIEW

**Comment by sbezverk** [comment]
@lpabon if you decide to go with it, it should be merged after @davidz627's PR.

**Comment by davidz627** [comment]
Is this a copy paste from the {code} repo? Any licensing issues with this...?

**Comment by sbezverk** [comment]
@davidz627 it is not just copy paste and as per @lpabon I added AUTHORS file with TheCodeTeam as authors. Do you still think there might be some issues?

**Comment by davidz627** [comment]
No worries! Was just making sure we had considered that.

Are there any other places/repos where we might want to use just the updated version of the mock driver? If so, maybe it would be better to just create a separate fork of the mock driver repo and update to v0.2.0 there. Then it is easier for others to have a dependency on that fork, and we can continue to attempt to send the changes "upstream." 

Alternatively, it might make sense to add them to this repo. Then this would become the one-stop-shop for driver testing and verification.

I'm not familiar enough with open source to say for sure which is the best option. I personally like adding them to this repo. @lpabon, @sbezverk , thoughts?

**Comment by sbezverk** [comment]
@lpabon @davidz627 I had to implement some additional changes as it seemed there was error code mismatch between the spec and e2e tests, as well as I extended a bit sanity checks inside of the mock driver. Now all e2e tests run successfully.

File: pkg/sanity/controller.go [hunk 1]
```diff
@@ -269,16 +269,12 @@ var _ = Describe("DeleteVolume [Controller Server]", func() {
 		}
 	})
 
-	It("should fail when no volume id is provided", func() {
+	It("should not fail when no volume id is provided", func() {
```
Comment (davidz627) [inline pkg/sanity/controller.go]: I don't think this is the correct behavior. See: https://github.com/container-storage-interface/spec/blob/master/spec.md#error-scheme

File: pkg/sanity/controller.go [hunk 2]
```diff
@@ -346,7 +342,7 @@ var _ = Describe("ValidateVolumeCapabilities [Controller Server]", func() {
 
 		serverError, ok := status.FromError(err)
 		Expect(ok).To(BeTrue())
-		Expect(serverError.Code()).To(Equal(codes.InvalidArgument))
+		Expect(serverError.Code()).To(Equal(codes.NotFound))
```
Comment (davidz627) [inline pkg/sanity/controller.go]: https://github.com/container-storage-interface/spec/blob/master/spec.md#error-scheme

same comment for many of these tests. Looks like if the argument itself is missing it should be "InvalidArgument" in the case that vol_id is provided but there is no volume found with that name, that is "NotFound"

**Comment by davidz627** [comment]
Looks great Serguei!

I think the main thing now is finding a place for this to live. It doesn't seem to be a Kubernetes specific library so I dont think that "kubernetes-csi/csi-test" is the right place to put it, the same argument could be made for csi-sanity. Both should be in the "csi" org proper.

We should bring it up at the next appropriate meeting

File: pkg/sanity/controller.go (hunk 2 above)
Comment (sbezverk) [reply pkg/sanity/controller.go]: @davidz627 I agree there is still a bit inconsistency. I have not validated all error codes returned/expected/in the spec. At one point it must be done. I was trying to match to the code returned by mock driver.

Comment (davidz627) [reply pkg/sanity/controller.go]: I just mean that since you are modifying the mock driver, I believe the Mock Driver should be changed, not the tests.

**Comment by lpabon** [comment]
@davidz627 afaik, this will most likely move to the same repo as the spec

File: pkg/sanity/controller.go (hunk 1 above)
Comment (lpabon) [reply pkg/sanity/controller.go]: I agree, I thought that it should fail.

Comment (sbezverk) [reply pkg/sanity/controller.go]: Gents, please check this part: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
I guess there are to points here: 
1. When volume id is not provided
2. When volume id does not exist

So for the first one, we should return codes.InvalidArgument
but for the second one it should be OK

Would you agree?

Comment (lpabon) [reply pkg/sanity/controller.go]: @sbezverk You are correct for DeleteVolume.

Comment (davidz627) [reply pkg/sanity/controller.go]: I see, looks like the spec is double specifying behavior here. Looks like your code is good. Agree with @lpabon we should fix the spec soon.

**Pull request merged**